scrapy/downloadermiddlewares/stats.py

```diff
@@ -24,7 +23,7 @@ def process_request(self, request, spider):
     def process_response(self, request, response, spider):
         self.stats.inc_value('downloader/response_count', spider=spider)
         self.stats.inc_value(f'downloader/response_status_count/{response.status}', spider=spider)
-        reslen = len(response_httprepr(response))
+        reslen = len(response.body)
```

Comment:
This seems to no longer take into account headers and such. I think the right solution is to create an alternative to `response_httprepr` that, without creating a new bytes object, providers the length of all parts involved.

Reply:
@Gallaecio 
`headers.to_string` used inside `response_httprepr` also duplicates headers in memory just like response_httprepr duplicates response.body
https://github.com/scrapy/scrapy/blob/26836c4e1ae9588ee173c5977fc6611364ca7cc7/scrapy/utils/response.py#L57-L60

https://github.com/scrapy/scrapy/blob/26836c4e1ae9588ee173c5977fc6611364ca7cc7/scrapy/http/headers.py#L75-L76

I think that each base http structure (request, response, headers, cookies etc.) needs to have it's own `get_size` method that calculate it's own size without duplicating it in memory as  raw `str` representation.

Reply:
I think we basically need to replace all those functions that generate strings with functions that calculate sizes, yes.